momadm_benchmarks/envs/congestion_game/congestion_game.py

```diff
@@ -0,0 +1,404 @@
+    - possible_agents
+    - action_spaces
+    - observation_spaces
+    These attributes should not be changed after initialization.
+    """
+
+    def __init__(
+        self,
+        problem_name="Braess_1_4200_10_c1",
+        num_agents=4200,
+        toll_mode="mct",
+        random_toll_percentage=0.1,
+        num_timesteps=1,
+        render_mode=None,
+    ):
+        """Initializes the congestion game.
+
+        Args:
+            problem_name: the name of the network that will be used
+            num_agents: number of agents in the network
+            toll_mode: the tolling mode that is used, tolls are either placed randomly "random" or using marginal cost tolling "mct"
+            random_toll_percentage: in the case of random tolling the percentage of roads that will be taxed
+            num_timesteps: number of timesteps (stateless, therefore always 1 timestep)
+            render_mode: render mode
+        """
+        # Read in the problem from the corresponding .json file in the networks directory
+        self.graph, self.od, self.routes = self._read_problem(problem_name)
+        # Keep track of the current flow on each link the network
+        self.flows = {f"{edge[0]}-{edge[1]}": 0 for edge in self.graph.edges}
+
+        # Episodes/Timesteps
+        self.num_timesteps = num_timesteps
+        self.episode_num = 0
+
+        self.render_mode = render_mode
+        self.possible_agents = ["agent_" + str(i) for i in range(num_agents)]
+        self.agents = self.possible_agents[:]
+        # each driver gets assigned a random origin-destination (OD) pair by _init_state()
+        self.drivers_od = self._init_state()
+        self.terminations = {agent: False for agent in self.agents}
+        self.truncations = {agent: False for agent in self.agents}
+
+        # compute the possible routes each agent can take based on its OD pair
+        route_choices_per_agent = [Discrete(len(self.routes[od])) for od in self.drivers_od]
+        # action space can be different for agents when there are multiple OD pairs with different numbers of routes
+        self.action_spaces = dict(zip(self.agents, route_choices_per_agent))
+        # stateless setting, agents receive a constant '0' as an observation in each timestep
+        self.observation_spaces = dict(
+            zip(
+                self.agents,
+                [
+                    Box(
+                        low=0,
+                        high=self.num_agents,
+                        shape=(1,),
+                        dtype=np.float32,
+                    )
+                ]
+                * num_agents,
+            )
+        )
+        # keep track of the maximum link latency and cost to scale the rewards returned to the agents
+        self._max_link_latency = None
+        self._max_link_cost = None
+        self.reward_spaces = dict(zip(self.agents, [Box(low=0, high=2000, shape=(2,))] * num_agents))
```

Comment:
Bounds should now be correct. The reward (actually a penalty) of each agent is the cost of the road it has chosen which itself is the sum of the different links of the road. The cost of the links are themselves scaled (between 0 and 1) so the reward should be at most equal to the length of the longest road (but negative since it is a penalty).